docs/proposals/08-shoot-additional-container-runtimes.md

```diff
@@ -0,0 +1,159 @@
+1. Hook the runtime binary into the containerd configuration file, so that the runtime becomes available to containerd.
+1. Apply a label to each node that allows identifying nodes where the runtime is available.
+1. Apply the relevant `RuntimeClass` to the Gardener cluster, to expose the functionality to users.
+1. Provide a `ValidatingWebhook` to catch invalid configurations. For example, Kata Containers on AWS requires a `machineType` of `i3.metal`.
+
+Since each operating system distribution has different methods of installing software (apt on ubuntu, zypper on SuSE, Torcx on CoreOS/Flatcar), we will enhance the operating system extensions to add a standard `installSoftware` script which can abstract the mechanics of installing new software on a node. The new runtime extensions will use this new functionality to install the relevant runtimes to the worker nodes.
+
+## Design Details
+
+1. In order to configure each worker machine in the cluster to work with CRI, the following configurations should be done:
+    1. Add kubelet execution flags:
+        1. --container-runtime=remote
+        2. --container-runtime-endpoint=unix:///.../containerd.socket
+    2. Make sure that containerd configuration file exist in path /etc/containerd/config.toml and it is configured to enable the CRI plugin (In the containerd config.toml "cri" plugin must not be disabled).
+    
+2. Containerd socket path and containerd configurations are different for each OS. To make sure the configurations above are set in each worker machine, each OS extension would be responsible to configure them.
+    1. os-ubuntu - 
+        1. Add a controlplane webhook to manipulate the OperatingSystemConfig and add the following flags to the kubelet execution command:
+            1. --container-runtime=remote
+            2. --container-runtime-endpoint=unix:///run/containerd/containerd.sock
+        2. Create a containerd configuration file: /etc/containerd/config.toml based on the current configuration file in the path: __TBD__ and remove the CRI disabled plugin.
+    2. os-coreos - 
+        1. Add a controlplane webhook to manipulate the OperatingSystemConfig and add the following flags to the kubelet execution command:
+           1. --container-runtime=remote
+           2. --container-runtime-endpoint=unix:///var/run/docker/libcontainerd/docker-containerd.socket
+        2. Create a containerd configuration file: /etc/containerd/config.toml based on the default configuration of containerd. 
+        3. Add a generator that will override the default containerd.service unit. It will create a new unit in: /etc/systemd/system/containerd.service unit that will start the containerd service with a configuration file path: /etc/containerd/config.toml.
+        
+    3. os-suse-jeos - __TBD__ 
+    
+3. Docker-monitor daemon and rotate log should be replaced with equivalent conatinerd services. __TBD__ 
```

Comment:
Do you already have an idea how this would look like? cc @ialidzhikov

Reply:
Not yet.